app/src/main/java/org/kiwix/kiwixmobile/zim_manager/local_file_transfer/DeviceListFragment.java

```diff
@@ -0,0 +1,364 @@
+  @Override
+  public void onPeersAvailable(WifiP2pDeviceList peers) {
+    searchingPeersProgressBar.setVisibility(View.GONE);
+    listViewPeerDevices.setVisibility(View.VISIBLE);
+
+    peerDevices.clear();
+    peerDevices.addAll(peers.getDeviceList());
+    ((WifiPeerListAdapter) getListAdapter()).notifyDataSetChanged();
+
+    if (peerDevices.size() == 0) {
+      Log.d(LocalFileTransferActivity.TAG, "No devices found");
+    }
+  }
+
+  public void clearPeers() {
+    peerDevices.clear();
+    ((WifiPeerListAdapter) getListAdapter()).notifyDataSetChanged();
+  }
+
+  void cancelAsyncTasks() {
+    if (peerGroupHandshakeAsyncTask != null) {
+      peerGroupHandshakeAsyncTask.cancel(true);
+    }
+
+    if (senderDeviceAsyncTaskArray != null) {
+      for (SenderDeviceAsyncTask task : senderDeviceAsyncTaskArray) {
+        task.cancel(true);
+      }
+    } else if (receiverDeviceAsyncTask != null) {
+      receiverDeviceAsyncTask.cancel(true);
+    }
+  }
+
+  public void onInitiateDiscovery() { // Setup UI for searching peers
+    searchingPeersProgressBar.setVisibility(View.VISIBLE);
+    listViewPeerDevices.setVisibility(View.INVISIBLE);
+    textViewPeerDevices.setVisibility(View.INVISIBLE);
+  }
+
+  /* From WifiP2pManager.ConnectionInfoListener callback-interface */
+  @Override
+  public void onConnectionInfoAvailable(WifiP2pInfo info) {
+    /* Devices have successfully connected, and 'info' holds information about the wifi p2p group formed */
+    groupInfo = info;
+    // Start handshake between the devices
+    peerGroupHandshakeAsyncTask = new PeerGroupHandshakeAsyncTask(this, groupInfo);
+    peerGroupHandshakeAsyncTask.execute();
+  }
+
+  public void setClientAddress(InetAddress clientAddress) {
+    if (clientAddress == null) {
+      // null is returned only in case of a failed handshake
+      showToast(localFileTransferActivity, R.string.device_not_cooperating, Toast.LENGTH_LONG);
+      localFileTransferActivity.closeLocalFileTransferActivity();
+      return;
+    }
+
+    // If control reaches here, means handshake was successful
+    selectedPeerDeviceInetAddress = clientAddress;
+    startFileTransfer();
+  }
+
+  private void startFileTransfer() {
+    fileTransferStarted = true;
+
+    if (groupInfo.groupFormed && !fileSender) {
+      displayTransferProgressFragment();
+
+      receiverDeviceAsyncTask = new ReceiverDeviceAsyncTask(this, transferProgressFragment);
+      receiverDeviceAsyncTask.execute();
+    } else if (groupInfo.groupFormed) {
+      {
+        Log.d(LocalFileTransferActivity.TAG, "Starting file transfer");
+
+        fileReceiverDeviceAddress =
+            (groupInfo.isGroupOwner) ? selectedPeerDeviceInetAddress : groupInfo.groupOwnerAddress;
+
+        // Hack for allowing slower receiver devices to setup server before sender device requests to connect
+        showToast(localFileTransferActivity, R.string.preparing_files, Toast.LENGTH_LONG);
+        for (int i = 0; i < 20000000; i++) ;
```

Comment:
Definitely not recommended. If you want to delay something you have a lot of options.
RxJava is very handy here there is a `delay` operator.
Start a new thread and post a runnable that sleeps that thread for a certain amount of time and then does work.
`Handler.post` schedules a runnable on that handlers thread on the next available frame.
`Handler.postDelayed` delay posting a runnable by X milliseconds.
`ExecutorService` more overblown but good for scheduling a repeating task.

Reply:
And really these are just bad options, what you really want is a callback when the server is setup which might require some polling? I don't know what errors you were getting

Reply:
Basically the file sender async task was trying to connect before the server (file receiver) had begun to accept connections.
Since the requirement is only to wait until the setup is ready, I figured this hack should do.
How about I repeat requests in the async-task until accepted, and use thread.sleep() for 1 sec in between each failed connection attempt?

Reply:
That is okay for me, for the moment anyway

Reply:
I'll get to it

Reply:
I've added a 3 second delay using thread.sleep() for now (the delay on the slower device used to be less than 1 second, so there's enough cushion without a noticeable lag) - [Commit](https://github.com/kiwix/kiwix-android/pull/1276/commits/ce8e3d3136021f2766e7ae71056ade01e6c5b080)

I'll use the methods you suggested once I get to replacing the async-tasks with Rx